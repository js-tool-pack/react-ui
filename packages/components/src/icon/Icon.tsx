import { getClassNames } from '@tool-pack/basic';
import { getComponentClass } from '@pkg/shared';
import { IconProps } from './icon.types';
import React from 'react';

const rootClass = getComponentClass('icon');
export const Icon: React.FC<IconProps> = React.forwardRef<
  HTMLElement,
  IconProps
>((props, ref) => {
  const { attrs = {}, className, children, color, size } = props;
  const { role = 'img', style = {} } = attrs;

  const fontSize =
    size === undefined
      ? style.fontSize
      : typeof size === 'number'
      ? size + 'px'
      : size;

  return (
    <i
      {...attrs}
      className={getClassNames(rootClass, attrs.className, className)}
      style={{ ...style, color: color || style.color, fontSize }}
      role={role}
      ref={ref}
    >
      {children ?? attrs.children}
    </i>
  );
});
