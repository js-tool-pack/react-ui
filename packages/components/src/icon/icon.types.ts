import React from 'react';
import { PropsBase } from '@pkg/shared';

type Attrs = React.HTMLAttributes<HTMLElement>;
export interface IconProps extends PropsBase {
  className?: Attrs['className'];
  size?: number | string;
  color?: React.CSSProperties['color'];
}
