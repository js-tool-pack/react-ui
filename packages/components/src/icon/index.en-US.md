---
category: Components
title: Icon
atomId: Icon
demo:
  cols: 2
group:
  title: Common
---

图标 Icon。

## 代码演示

<style>
.demo-icon-icons {font-size: 30px; color: var(--t-primary-bg-color)} 
.demo-icon-icons .t-icon {padding: 2px} 
.demo-icon-icons .t-icon:hover {
background: #dcdcdc;
cursor: pointer;
}
</style>

<!-- prettier-ignore -->
<code src="./demo/basic.tsx"></code>
<code src="./demo/icons.tsx"></code>

## API

图标的属性说明如下：

| 属性  | 说明                                                                              | 类型                                         | 默认值 | 版本 |
| ----- | --------------------------------------------------------------------------------- | -------------------------------------------- | ------ | ---- |
| size  | 图标大小，当传入的是 number 类型时，会加上`px`单位后缀；也可以直接使用 style 设置 | string \| number                             | `1em`  |      |
| color | 图标颜色，可以直接使用 style 设置                                                 | string                                       | false  |      |
| attrs | 组件 html 根元素的所有属性                                                        | Partial\<React.HTMLAttributes\<HTMLElement>> | --     | --   |
