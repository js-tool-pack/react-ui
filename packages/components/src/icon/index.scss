@use '../namespace' as Name;
@use 'sass:math';

.#{Name.$icon} {
  position: relative;
  display: inline-block;
  width: 1em;
  height: 1em;
  line-height: 1em;
  > svg {
    width: 1em;
    height: 1em;
  }
}

$offset: 187;
$duration: 2s;
$l: Name.$icon-loading;

.#{$l} {
  animation: loading-rotate2 $duration linear infinite;
  &__path {
    stroke-dasharray: $offset;
    stroke-dashoffset: 0;
    transform-origin: center;
    animation: loading-dash $duration ease-in-out infinite;
  }
}

@keyframes loading-rotate2 {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(270deg);
  }
}

@keyframes loading-dash {
  0% {
    stroke-dashoffset: $offset;
  }
  50% {
    stroke-dashoffset: math.div($offset, 4);
    transform: rotate(135deg);
  }
  100% {
    stroke-dashoffset: $offset;
    transform: rotate(450deg);
  }
}
