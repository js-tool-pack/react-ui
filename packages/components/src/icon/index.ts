export type { IconProps } from './icon.types';
export * from './Icon';
