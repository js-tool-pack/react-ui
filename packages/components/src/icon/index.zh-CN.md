---
category: Components
title: Icon
subtitle: 图标
demo:
  cols: 2
group:
  title: 通用
---

图标 Icon。

## 代码演示

<style>
.demo-icon-icons {font-size: 30px; color: var(--t-primary-bg-color)} 
.demo-icon-icons .t-icon {padding: 2px} 
.demo-icon-icons .t-icon:hover {
background: #dcdcdc;
cursor: pointer;
}
</style>

<!-- prettier-ignore -->
<code src="./demo/basic.tsx"></code>
<code src="./demo/icons.tsx"></code>

## API

图标的属性说明如下：

| 属性  | 说明                                                                              | 类型             | 默认值 | 版本 |
| ----- | --------------------------------------------------------------------------------- | ---------------- | ------ | ---- |
| size  | 图标大小，当传入的是 number 类型时，会加上`px`单位后缀；也可以直接使用 style 设置 | string \| number | `1em`  |      |
| color | 图标颜色，可以直接使用 style 设置                                                 | string           | false  |      |

支持原生 HTML 的其他所有属性。
