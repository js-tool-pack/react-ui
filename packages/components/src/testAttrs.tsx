import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
