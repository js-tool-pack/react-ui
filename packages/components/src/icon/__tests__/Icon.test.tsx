import { render } from '@testing-library/react';
import { testAttrs } from '~/testAttrs';
import { Icon } from '~/icon';

describe('Icon', () => {
  testAttrs(Icon);

  test('basic', () => {
    const { container } = render(<Icon size={20}>123</Icon>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('size', () => {
    const { container } = render(<Icon size={20}>123</Icon>);
    expect(container.firstChild).toHaveStyle({ fontSize: '20px' });
  });

  test('color', () => {
    const { container } = render(<Icon color="#fff">123</Icon>);
    expect(container.firstChild).toHaveStyle({ color: '#fff' });
  });

  test('className', () => {
    const { container } = render(<Icon className="foo">123</Icon>);
    expect(container.firstChild).toHaveClass('t-icon foo');
  });

  test('role', () => {
    const { container } = render(<Icon className="foo">123</Icon>);
    expect(container.firstChild).toHaveAttribute('role', 'img');
  });
});
