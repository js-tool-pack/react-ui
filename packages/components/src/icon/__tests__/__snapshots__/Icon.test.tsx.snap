// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon basic 1`] = `
<i
  class="t-icon"
  role="img"
  style="font-size: 20px;"
>
  123
</i>
`;
